src/SessionSetup/ServiceDeployManager.cpp

```diff
@@ -449,78 +451,102 @@ ErrorMessageOr<void> ServiceDeployManager::CopyOrbitUserSpaceInstrumentationLibr
-  // implemented in OrbitSshQt::Task.
   emit statusMessage("Starting OrbitService on the remote instance...");
 
-  std::string task_string = "sudo --stdin /tmp/OrbitService";
-  if (absl::GetFlag(FLAGS_devmode)) {
-    task_string += " --devmode";
-  }
+  std::string task_string = GenStartOrbitServiceCommand(deployment_config);
   orbit_service_task_.emplace(&session_.value(), task_string);
 
-  orbit_service_task_->Write(absl::StrFormat("%s\n", config.root_password));
+  if (std::holds_alternative<BareExecutableAndRootPasswordDeployment>(deployment_config)) {
+    const auto& config = std::get<BareExecutableAndRootPasswordDeployment>(deployment_config);
+    orbit_service_task_->Write(absl::StrFormat("%s\n", config.root_password));
+    // TODO(antonrohr) Check whether the password was incorrect.
+    // There are multiple ways of doing this. the best way is probably to have a
+    // second task running before OrbitService that sets the SUID bit. It might be
+    // necessary to close stdin by sending EOF, since sudo would ask for trying to
+    // enter the password again. Another option is to use std err as soon as its
+    // implemented in OrbitSshQt::Task.
+  }
 
   orbit_qt_utils::EventLoop loop{};
   auto error_handler =
       ConnectErrorHandler(&loop, &orbit_service_task_.value(), &orbit_ssh_qt::Task::errorOccurred);
-  auto quit_handler =
-      ConnectQuitHandler(&loop, &orbit_service_task_.value(), &orbit_ssh_qt::Task::started);
   auto cancel_handler = ConnectCancelHandler(&loop, this);
+  QObject::connect(
+      &orbit_service_task_.value(), &orbit_ssh_qt::Task::finished, &loop, [&loop](int exit_code) {
+        // TODO(http://b/221369463): So far this is only reporting a rather generic exit code. This
+        // will be extended in a subsequent PR.
+        loop.error(ErrorMessage{
+            absl::StrFormat("The service exited prematurely with exit code %d.", exit_code)});
+      });
+
+  std::string stdout_buffer;
+
+  QObject::connect(
+      &orbit_service_task_.value(), &orbit_ssh_qt::Task::readyReadStdOut, &loop, [&]() {
+        stdout_buffer.append(orbit_service_task_->ReadStdOut());
+
+        // That's what we expect the service to send through stdout when it's ready to
+        // accept a connection from the client.
+        constexpr std::string_view kReadyKeyword = "READY";
+
+        if (absl::StrContains(stdout_buffer, kReadyKeyword)) {
+          ORBIT_LOG("The service reported to be ready to accept connections.");
+          loop.quit();
+          return;
+        }
+
+        // This is protecting us against consuming unreasonable amount of memory when
+        // for whatever reason there is a lot of data coming through the stdout
+        // channel.
+        constexpr size_t kMaxBufferSize = 100 * 1024;  // 100 KiB
+
+        if (stdout_buffer.size() > kMaxBufferSize) {
+          constexpr size_t kNumberOfBytesToRemove = kMaxBufferSize - kReadyKeyword.size();
```

Comment:
I'm confused here. The way this is written, it sounds like the _remaining_ size be ~`kMaxBufferSize`. Although I guess I'm missing the intent.

Reply:
Fixed/Changed/whatever. Should be better now!